src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/MemoryBlock.cs

```diff
@@ -315,33 +314,23 @@ internal string PeekUtf8NullTerminated(int offset, byte[]? prefix, MetadataStrin
         /// If a value other than '\0' is passed we still stop at the null terminator if encountered first.</param>
         /// <param name="numberOfBytesRead">The number of bytes read, which includes the terminator if we did not hit the end of the block.</param>
         /// <returns>Length (byte count) not including terminator.</returns>
-        internal int GetUtf8NullTerminatedLength(int offset, out int numberOfBytesRead, char terminator = '\0')
+        internal int GetUtf8NullTerminatedLength(int offset, out int numberOfBytesRead, char terminator)
         {
             CheckBounds(offset, 0);
 
             Debug.Assert(terminator <= 0x7f);
 
-            byte* start = Pointer + offset;
-            byte* end = Pointer + Length;
-            byte* current = start;
-
-            while (current < end)
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(Pointer + offset, Length - offset);
+            int length = terminator != '\0' ?
+                span.IndexOfAny((byte)0, (byte)terminator) :
```

Comment:
Out-of-bounds read if `offset` is negative?

Reply:
Yep, `CheckBounds(offset, 0)` only validates against `Length`, and not `Pointer`.

Reply:
Isn't that already the case?

Reply:
I see this is in fact validated in `CheckBounds` with some casting tricks:

```csharp
if (unchecked((ulong)(uint)offset + (uint)byteCount) > (ulong)Length)
```